overlord/fdestate/dbx.go

```diff
@@ -0,0 +1,457 @@
+	//
+	// TODO flip operation state to Doing and conflict on new request?
+
+	if op.Status != state.DoStatus {
+		return &snapstate.ChangeConflictError{
+			ChangeKind: "fde-efi-secureboot-db-update",
+			Message:    "cannot perform DBX update 'startup' action when conflicting actions are in progress",
+		}
+	}
+
+	op.SetStatus(state.DoingStatus)
+	if err := updateExternalOperation(st, op); err != nil {
+		return nil
+	}
+	// TODO what to do with status if resealing fails?
+
+	err = func() error {
+		st.Unlock()
+		defer st.Lock()
+
+		return postUpdateReseal(st, method, "efi-secureboot-update-startup")
+	}()
+	if err != nil {
+		return fmt.Errorf("cannot complete post update reseal in startup action: %w", err)
+	}
+
+	return abortEFISecurebootDBUpdateChange(st, op,
+		"startup action invoked while a change is in progress")
+}
+
+type EFISecurebootKeyDatabase int
+
+const (
+	EFISecurebootPK EFISecurebootKeyDatabase = iota
+	EFISecurebootKEK
+	EFISecurebootDB
+	EFISecurebootDBX
+)
+
+// EFISecureBootDBUpdatePrepare notifies notifies that the local EFI key
+// database manager is about to update the database.
+func EFISecureBootDBUpdatePrepare(st *state.State, db EFISecurebootKeyDatabase, payload []byte) error {
+	method, err := device.SealedKeysMethod(dirs.GlobalRootDir)
+	if err != nil {
+		if err == device.ErrNoSealedKeys {
+			return nil
+		}
+		return err
+	}
+
+	st.Lock()
+	defer st.Unlock()
+
+	op, err := findPendingExternalOperationByKind(st, "fde-efi-secureboot-db-update")
+	if err != nil {
+		return err
+	}
+
+	if op != nil {
+		return &snapstate.ChangeConflictError{
+			ChangeKind: "fde-efi-secureboot-db-update",
+			Message:    "cannot start a new DBX update when conflicting actions are in progress",
+		}
+	}
+
+	// TODO check that there are no ongoing changes for base/kernel/gadget
+
+	op, err = addEFISecurebootDBUpdateChange(st, payload)
+	if err != nil {
+		return err
+	}
+
+	// at this point, after unlocking, operation will show up in the state, and
+	// any calls to perform it again will raise a conflict, even if performed
+	// right after briefly unlocking the state for reseal
+
+	err = func() error {
+		st.Unlock()
+		defer st.Lock()
+
+		return boot.WithBootChains(func(bc *boot.ResealKeyForBootChainsParams) error {
+			logger.Debugf("attempting reseal for DBX update")
+			logger.Debugf("boot chains: %v\n", bc)
+			logger.Debugf("DBX update payload: %x", payload)
+
+			// TODO use a different helper for resealing
+			return backendResealKeysForSignaturesDBUpdate(
+				unlockedStateUpdater(st, "efi-secureboot-update-db-prepare"),
+				method, dirs.GlobalRootDir, bc, payload,
+			)
+		}, method)
+	}()
+	if err != nil {
+		reason := fmt.Sprintf("initial reseal failed: %v", err)
+		if errAbort := abortEFISecurebootDBUpdateChange(st, op, reason); errAbort != nil {
+			logger.Noticef("cannot abort pending FDE change: %v", err)
+		}
+		// TODO return error
+		return fmt.Errorf("cannot reseal keys for DBX update: %w", err)
+	}
```

Comment:
The reseal with prepare data should be moved to `*-prepare` task, then in prepare's undo we can reseal again but with the current state of DBX, thus keeping symmetry
At the same time the `*Prepare` call will now need t synchronize with the `*-prepare` task completion